Perform optimistic reads while filling gaps in CachingFileSystem

carlopi opened the pull request:
Example situation, you have read the following byte ranges: `(1000000, 1300000) and (2000000, 2300000)`

And you are about to read the (uncached) range `(800000, 9500000)`, that would generate the following covered ranges: `(800000, 9500000), (1000000, 1300000) and (2000000, 2300000)`

Now the idea of the PR is: why not read `(800000, 1000000)` instead of `(800000, 9500000)` ?
It's slightly bigger but will generate the following ranges: `(800000, 1300000) and (2000000, 2300000)`, that might avoid an extra range request down the line if we ever discover we need (some of) the bytes in `(9500000, 1000000)` to be read.

Basically this is swapping doing bigger ranges, that means more memory pressure AND somewhat more inefficient requests in exchange from the uncertain prospect of skipping a whole request at some future point in time, with the assumption that extra requests are more costly that bigger requests, and that chances are that even in the same query you might end up reading the rest of the ranges.

Idea is similar (but more limited and so safer) to buffering file system in DuckDB-Wasm, where we trade some extra work on current requests hoping for that to be offset by skipping a full request at some later point.

From my basic tests this is very much dependent on data, for example running `PRAGMA tpch(1)` at SF=10 on a datalake (so storage are parquet files) cuts from 50 range requests to 38, for basically the same total transfered bytes. But then running `PRAGMA tpch(1)` to `PRAGMA tpch(22)` (with the same cache) shows no advantages in other queries, and 1% extra total transfered bytes.

Change is small, and seems to me to make sense on paper (or does it?), but some more solid numbers would be handy.

Open questions:
1. what should be the exact formula to decide whether to fetch extra bytes or not. Currently it's whether [location, end] range needs to be expanded by less than (location-end) bytes (so guaranteeing worse case of 2x) AND by less than 1MB.
2. whether there should be a knob/setting to regulate behaviour
3. validating this idea on other data sets

carlopi committed:
Perform optimistic reads while filling gaps in CachingFileSystem

Example situation, you have read the following byte ranges:
1000000-1300000 and 2000000-2300000

And you are about to read the (uncached) range 800000-9500000, that would generate the following covered ranges:
800000-9500000, 1000000-1300000 and 2000000-2300000

Now the idea of the PR is: why not read 800000-1000000 instead of 800000-9500000 ?
It's slightly bigger but will generate the following ranges: 800000-1300000 and 2000000-2300000, that might avoid an extra range request down the line.

Basically this is swapping doing bigger ranges, that means more memory pressure AND somewhat more inefficient requests in exchange from the uncertain prospect of skipping a whole request at some future point in time.

Idea is similar (but more limited and so safer) to buffering file system in DuckDB-Wasm, where we trade some extra work on current requests hoping for that to be offset by skipping a full request at some later point.

From my basic tests this is very much dependent on data, for example running `PRAGMA tpch(1)` at SF=10 on a datalake (so storage are parquet files) cuts from 50 range requests to 38, for basically the same total transfered bytes.
But then running `PRAGMA tpch(1)` to `PRAGMA tpch(22)` (with the same cache) shows no advantages in other queries, and 1% extra total transfered bytes.

Change is small, and seems to me to make sense on paper (or does it?), but some more solid numbers would be handy.

Open questions:
1. what should be the exact formula to decide whether to fetch extra bytes or not. Currently it's whether [location, end] range needs to be expanded by less than (location-end) bytes (so guaranteeing worse case of 2x) AND by less than 1MB.
2. whether there should be a knob/setting to regulate behaviour
3. validating this idea on other data sets

lnkuiper commented on src/storage/caching_file_system.cpp:
```diff
@@ -90,30 +90,42 @@ BufferHandle CachingFileHandle::Read(data_ptr_t &buffer, const idx_t nr_bytes, c
+			// Grow the range from location to start_location_of_next_range, so that to fill gaps in the cached ranges
+			new_nr_bytes = start_location_of_next_range - location;
+		}
+	}
```
Let's make `start_location_of_next_range` an `optional_idx`, and `nr_bytes` non-`const`, then we can do something like this:
```C++
if (start_location_of_next_range.IsValid()) {
	static constexpr idx_t MINIMUM_ADDED_BYTES = 10000;
	static constexpr idx_t MAXIMUM_ADDED_BYTES = 1000000;

	const auto added_bytes_bound = ClampValue(MINIMUM_ADDED_BYTES, nr_bytes, MAXIMUM_ADDED_BYTES);
	const auto new_nr_bytes = start_location_of_next_range - location;
	if (new_nr_bytes <= nr_bytes + added_bytes_bound) {
		nr_bytes = new_nr_bytes;
	}
}
```

carlopi replied:
Thanks! I will fix this up and undraft

lnkuiper replied:
Discussed over Discord but adding here so we don't forget: `start_location_of_next_range` can be initialized with the file size (if available)

lnkuiper reviewed (changes requested):
I really like this idea! I have left a comment below with a (readability) improvement to the logic around `new_nr_bytes`. I'm not sure what the formula should be exactly, but if this seems OK in your experiment then I'm OK with it

carlopi committed:
filling gaps in CachingFileSystem: implement feedback

carlopi commented:
I implemented the feedback, but for having `start_location_of_next_range` initialised to the size of the file.

I am not sure on that one, might be worth, but given parquet files do store metadata there I am not sure it's net positive, but it can be looked at and implemented in a follow up.

lnkuiper commented on src/storage/caching_file_system.cpp:
```diff
@@ -79,6 +79,21 @@ FileHandle &CachingFileHandle::GetFileHandle() {
 }
 
+static bool ShouldExpandToFillGap(const idx_t current_length, const idx_t added_length) {
+	const idx_t MAX_BOUND_TO_BE_ADDED_LENGTH = 1048576;
```
Don't we want to have a minimum bound of bytes that is always OK to add?

For example, if we're reading 500 bytes, this will only allow 500 bytes to be added. I think in most cases it's OK to read 4/8/16k. We might be able to save a few requests that way.

It could also potentially be interesting to implement this idea as a "minimum read size" instead, that is generally applicable (regardless of whether it completely fills the gap to the next byte range). Reads that go through `CachingFileSystem` could be increased to be at least 4/8k.

I think this would be great for Parquet files that are highly compressed. Some Parquet files that compress well have page sizes of just a few hundred bytes. Optimistically fetching a few kb wouldn't cause much overhead, but could save us a few roundtrips.

lnkuiper reviewed (commented):
Thanks for the changes! I have one thought, that doesn't necessarily need to be implemented in this PR, I just think it could be interesting:

lnkuiper reviewed (approved):
This is good to go from my side, we have an open issue for further improvements.
